src/mito2/src/memtable/time_partition.rs

```diff
@@ -377,6 +499,67 @@ impl TimePartitions {
         inner.parts.clone()
     }
 
+    /// Find existing partitions that match the bulk data's time range and identify
+    /// any new partitions that need to be created
+    fn find_partitions_by_time_range<'a>(
+        &self,
+        existing_parts: &'a [TimePartition],
+        min: Timestamp,
+        max: Timestamp,
+    ) -> Result<(Vec<&'a TimePartition>, Vec<Timestamp>)> {
+        let mut matching = Vec::new();
+
+        let mut present = HashSet::new();
+        // First find any existing partitions that overlap
+        for part in existing_parts {
+            let Some(part_time_range) = part.time_range.as_ref() else {
+                matching.push(part);
+                return Ok((matching, Vec::new()));
+            };
+
+            if !(max < part_time_range.min_timestamp || min >= part_time_range.max_timestamp) {
+                matching.push(part);
+                present.insert(part_time_range.min_timestamp.value());
+            }
+        }
+
+        // safety: self.part_duration can only be present when reach here.
+        let part_duration = self.part_duration.unwrap();
+        let timestamp_type = self
+            .metadata
+            .time_index_column()
+            .column_schema
+            .data_type
+            .as_timestamp()
+            .unwrap(); // SAFETY: timestamp column must be a timestamp.
+
+        let part_duration_sec = part_duration.as_secs() as i64;
+        // SAFETY: Timestamps won't overflow when converting to Second.
+        let bulk_start_sec = min
+            .convert_to(TimeUnit::Second)
+            .unwrap()
+            .value()
+            .div_euclid(part_duration_sec);
+        let bulk_end_sec = max
+            .convert_to(TimeUnit::Second)
+            .unwrap()
+            .value()
+            .div_euclid(part_duration_sec);
+
+        let missing: Vec<_> = (bulk_start_sec..=bulk_end_sec)
```

Comment:
If the input data is sparse, and the end second is greater than the start second a lot, I wonder this loop may be costly. We may check the bucket num and the timestamp num and use the smaller one to compute the bucket num.

Reply:
> If the input data is sparse, and the end second is greater than the start second a lot, I wonder this loop may be costly. We may check the bucket num and the timestamp num and use the smaller one to compute the bucket num.

Here the variable name is a little bit confusing. `bulk_start_sec` and `bulk_end_sec` are not seconds, but actually the bulk time bucket index because we it's assigned to `batch.max_ts / part_duration`...

Reply:
I think the issue still exists. For example, if the time bucket is 1h, and I insert the following timestamps:
```
[2022-01-01T00:00:00, 2023-01-01T00:00:00, 2024-01-01T00:00:00]
```
If we use the min/max to iterate the buckets, we will iterate over all buckets hour by hour between 2022 and 2024.